src/System.Memory/src/System/Buffers/Text/Utf8Formatter/Utf8Formatter.Integer.Unsigned.Default.cs

```diff
@@ -0,0 +1,44 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Runtime.CompilerServices;
+
+namespace System.Buffers.Text
+{
+    /// <summary>
+    /// Methods to format common data types as Utf8 strings.
+    /// </summary>
+    public static partial class Utf8Formatter
+    {
+        private static bool TryFormatUInt64Default(ulong value, Span<byte> buffer, out int bytesWritten)
+        {
+            ref byte utf8Bytes = ref buffer.DangerousGetPinnableReference();
+
+            ulong left = value;
+            for (int i = 0; i < buffer.Length; i++)
```

Comment:
Have you tried to compute the required space upfront instead of doing the swapping? E.g. something like:

```
        int digits = 1;

        uint part;
        if (value >= 10000000)
        {
            if (value >= 100000000000000)
            {
                part = (uint)(value % 100000000000000);
                digits += 14;
            }
            else
            {
                part = (uint)(value % 10000000);
                digits += 7;
            }
        }
        else
        {
            part = (uint)value;
        }

        if (part < 10) { }
        else if (part < 100) digits += 1;
        else if (part < 1000) digits += 2;
        else if (part < 10000) digits += 3;
        else if (part < 100000) digits += 4;
        else if (part < 1000000) digits += 5;
        else if (part < 10000000) digits += 6;
```

Reply:
> Have you tried to compute the required space upfront instead of doing the swapping?

That is what some of the other implementations do (call `FormattingHelpers.CountDigits` up front). But they do it as a loop (where the value is divided by 10 over and over). I will play around with the approach you outlined.
https://github.com/dotnet/corefx/blob/master/src/System.Memory/src/System/Buffers/Text/Utf8Formatter/Utf8Formatter.Integer.Signed.D.cs#L17

Reply:
BTW: `if (part < 10000000)` on the last line in my code fragment should replaced by Assert - it should be always true.

Reply:
Should these be modulo or division?
`part = (uint)(value % 100000000000000);`
...
`part = (uint)(value % 10000000);`

The way it is written atm, it doesn't give the correct result. For example:
```C#
ulong value = 1234567890
CountDigits(value) => 14 // should be 10.
```

Reply:
Right, these should be divisions.

Reply:
Computing the required space upfront for numbers with 3 or more digits is better. In most cases, we get a 5%+ gain. However, for single digits, there is a 15-20% regression.

![image](https://user-images.githubusercontent.com/6527137/33297149-5ce9d182-d394-11e7-8192-0055ae5fb61a.png)

Reply:
You should have dedicated path upfront for single digits, like:

```
if (value < 10)
{
    if (buffer.Length == 0)
         goto FalseReturn;
     utf8Bytes = (byte)('0' + (byte)value);
     return true;
}
```